### ISSUE TITLE
 libncurses.so.5 link disappeared from ncurses derivation after "staging" merge

### ISSUE THREAD

**Issue by peti** [NixOS/nixpkgs#7810]
The binary version of GHC 6.10.2 is broken in `master`. That compiler is used to bootstrap GHC 6.10.4 and 6.12.3, which means that those compilers (and their respective package sets) are now broken, too.

The error seems relatively harmless. The build succeeds, but the resulting GHC binary doesn't work. Citing from http://hydra.cryp.to/build/815012/log/raw:

> /nix/store/ir0ip3jiivcwcg4zdzf6rj616f4pnpr7-ghc-6.10.2-binary/lib/ghc-6.10.2/ghc: error while loading shared libraries: libncurses.so.5: cannot open shared object file: No such file or directory

**Comment by peti** [comment]
@wkennington, @ttuegel it appears this breakage was caused by your `ncurses` re-factoring. Prior to the `staging` merge, `ncurses` provided `libncurses.so.5`, but that path now appears to be gone, hence the broken `ghc` build.

**Comment by ttuegel** [comment]
@peti The ncurses refactoring is due to @wkennington. I had to bring back `libncurses.so.5` so that older GHCs would at least build; I don't know why it is found at build time but not runtime.

**Comment by vcunat** [comment]
Hmm, I looked on ncurses, and it confused me that the links have such a strange date. I thought it impossible, as nix is supposed to normalize them AFAIK, but even after ``--check`` I get the same:
```
$ ls -l ./result/lib
total 1852
-r--r--r-- 35 root root     18 Jan  1  1970 libcurses.so
-r--r--r-- 35 root root     18 Jan  1  1970 libcursesw.so
lrwxrwxrwx 33 root root     10 May 15  2012 libform.a -> libformw.a
-r--r--r-- 33 root root     15 Jan  1  1970 libform.so
-r--r--r--  2 root root 135124 Jan  1  1970 libformw.a
lrwxrwxrwx 64 root root     13 May 15  2012 libformw.so -> libformw.so.5
lrwxrwxrwx 32 root root     15 May 15  2012 libformw.so.5 -> libformw.so.5.9
-r-xr-xr-x  2 root root  79968 Jan  1  1970 libformw.so.5.9
lrwxrwxrwx 33 root root     10 May 15  2012 libmenu.a -> libmenuw.a
-r--r--r-- 33 root root     15 Jan  1  1970 libmenu.so
-r--r--r--  2 root root  69062 Jan  1  1970 libmenuw.a
lrwxrwxrwx 64 root root     13 May 15  2012 libmenuw.so -> libmenuw.so.5
lrwxrwxrwx 32 root root     15 May 15  2012 libmenuw.so.5 -> libmenuw.so.5.9
-r-xr-xr-x  2 root root  38273 Jan  1  1970 libmenuw.so.5.9
lrwxrwxrwx 33 root root     13 May 15  2012 libncurses.a -> libncursesw.a
lrwxrwxrwx  2 root root     15 Jan  1  1970 libncurses++.a -> libncurses++w.a
-r--r--r-- 35 root root     18 Jan  1  1970 libncurses.so                                                                                        
-r--r--r--  2 root root 167678 Jan  1  1970 libncurses++w.a                                                                                      
-r--r--r--  2 root root 811852 Jan  1  1970 libncursesw.a                                                                                        
lrwxrwxrwx 64 root root     16 May 15  2012 libncursesw.so -> libncursesw.so.5                                                                   
lrwxrwxrwx 32 root root     18 May 15  2012 libncursesw.so.5 -> libncursesw.so.5.9                                                               
-r-xr-xr-x  2 root root 456146 Jan  1  1970 libncursesw.so.5.9                                                                                   
lrwxrwxrwx 33 root root     11 May 15  2012 libpanel.a -> libpanelw.a                                                                            
-r--r--r-- 33 root root     16 Jan  1  1970 libpanel.so                                                                                          
-r--r--r-- 10 root root  27426 Jan  1  1970 libpanelw.a                                                                                          
lrwxrwxrwx 64 root root     14 May 15  2012 libpanelw.so -> libpanelw.so.5                                                                       
lrwxrwxrwx 32 root root     16 May 15  2012 libpanelw.so.5 -> libpanelw.so.5.9                                                                   
-r-xr-xr-x  2 root root  14100 Jan  1  1970 libpanelw.so.5.9                                                                                     
dr-xr-xr-x  1 root root    186 Jan  1  1970 pkgconfig                                                                                            
lrwxrwxrwx 33 root root     17 May 15  2012 terminfo -> ../share/terminfo
```

**Comment by Warbo** [comment]
I've run into a similar looking issue with building GHC 7.4.2 to bootstrap 7.8.4 and 7.10.1:

```
$ git status
HEAD detached at f93a8ee
nothing to commit, working directory clean
$ nix-env -f . -i -A haskellngPackages.cabal2nix
replacing old ‘cabal2nix-1.73’
installing ‘cabal2nix-20150505’
these derivations will be built:
  /nix/store/vawns4div0z5cq1w1wvch0jgw7zq40fj-ghc-7.4.2-binary.drv
  /nix/store/vdgwwldb3a8fwv9bvmrc21cqqmb0zidl-ghc-7.8.4.drv
  /nix/store/68pkhyfymp5vrbysyzgrsbd621gxvqyw-ghc-7.10.1.drv
  /nix/store/zynmnah9il96mbrv5zpn06h97iy4m32r-hscolour-1.23.drv
  /nix/store/xc4y1m05d7myhxh2ij0xzv59bpv8mv6i-haskell-mtl-2.2.1.drv
  /nix/store/0his2dczhr78iv4x4g7cybh0646nylcm-happy-1.19.5.drv
  /nix/store/b9kpanl3fsbi4zn2q22clxlai55v514g-jailbreak-cabal-1.3.drv
  /nix/store/bhj0zlmxq2gcc637qhzjjv5x565ba67k-haskell-random-1.1.drv
  /nix/store/dy0ljr7sqksgdz727gc304s2xhsblmr6-haskell-old-locale-1.0.0.7.drv
  /nix/store/5bdd0la8s2fcb4sb23n0bzipzhjbikx1-haskell-text-1.2.0.4.drv
  /nix/store/f3qx01vds5jy4hcfavb98wiwfh4rkjql-haskell-xml-1.3.14.drv
  /nix/store/iwana7rmh8sj6qdik8xi5nwqyg0pahkd-haskell-hostname-1.0.drv
  /nix/store/vjn7ln29k4b2vf4p6isyxys3274b7b2z-haskell-regex-base-0.93.2.drv
  /nix/store/kb4svawazq1gs3v15kj0s89h7ny7zs89-haskell-regex-posix-0.95.2.drv
  /nix/store/rcvy6w9lyhpxlki0qjxsmpv9sf0fngq8-haskell-ansi-terminal-0.6.2.1.drv
  /nix/store/wv9g9936wk7k31scxk11fzzvpg0w9dzs-haskell-ansi-wl-pprint-0.6.7.2.drv
  /nix/store/26w52jma2km5sr6v9n6s80ibdgnssq4g-haskell-test-framework-0.8.1.1.drv
  /nix/store/x415hvji84qvvii6w140rqf4xlwzl2rw-haskell-primitive-0.5.4.0.drv
  /nix/store/kkymzf9lw9dc3lx8qm2yi393p99gh89f-haskell-tf-random-0.5.drv
  /nix/store/hkidkdv6amxipy6l4ff3ya8q5kx34qzs-haskell-QuickCheck-2.7.6.drv
  /nix/store/xsx279k1zk1a6yvli4g714vc7smsl0ap-haskell-dlist-0.7.1.1.drv
  /nix/store/1ih4qxqdl8vqpi0dk13r7q8xhn0nj7sk-haskell-Glob-0.7.5.drv
  /nix/store/bwjismngsf0h8r0bs1yfm142fy3j5y9l-haskell-transformers-compat-0.4.0.4.drv
  /nix/store/24gl6jbgdw22h9jsqibz5839vwr50199-haskell-HUnit-1.2.5.2.drv
  /nix/store/dxis4x03gfhhy4xvjbr06ry6a299qkp5-haskell-extensible-exceptions-0.1.1.4.drv
  /nix/store/39x8w0wj93wbgq17qqm3yiwcnp1f1nw4-haskell-test-framework-quickcheck2-0.3.0.3.drv
  /nix/store/4b35bb5vd7ypajmpl7dkjv0w2fr81msw-haskell-test-framework-hunit-0.3.0.1.drv
  /nix/store/4shwbpv7f5sxa4hzbb0371cdk7bgy5lm-haskell-hashable-1.2.3.2.drv
  /nix/store/9lxd094rms6y9g940b41lrcv3rd62sp4-haskell-nats-1.drv
  /nix/store/khr6syzd9z96jgny94yhif3hs9nsrxyp-haskell-syb-0.4.4.drv
  /nix/store/16fphxchlm1zma0nhcj138m6vc1ckivv-haskell-ChasingBottoms-1.3.0.12.drv
  /nix/store/bandyp0iwcij2rs8gq056qj8398sk6f8-haskell-unordered-containers-0.2.5.1.drv
  /nix/store/m0z2pmkpy7ih14nwxf5qkv93ljfv69j1-haskell-semigroups-0.16.2.2.drv
  /nix/store/7y35byp558cqr2vlq5lfdki5bcydyv97-haskell-stm-2.4.4.drv
  /nix/store/nj9861y24kiq4x04zpkzchb3j01jf51i-haskell-StateVar-1.1.0.0.drv
  /nix/store/zybkx8c7xkp0g8vb48c0kv6vxik08w19-haskell-void-0.7.drv
  /nix/store/g6s2f77y94b1rwsngmi3mvx6rq1s2yvb-haskell-contravariant-1.3.1.1.drv
  /nix/store/rmjikkk1lli3w50jnmzbn8z9alw05j2p-haskell-nanospec-0.2.1.drv
  /nix/store/2d06n3zpadrbd3ijm0g182kagzfkyfqi-haskell-silently-1.2.4.1.drv
  /nix/store/c4qaais955ddg5fj7552kmigi8zv29vi-haskell-hspec-expectations-0.6.1.1.drv
  /nix/store/2b8xqqi2mq4m31psr2ijx77j2zcgbcq1-haskell-quickcheck-io-0.1.1.drv
  /nix/store/dvvdkhq76031rk95wh2c3qmf123xqwmd-haskell-async-2.0.2.drv
  /nix/store/jvqgmn8qszrhm95b8is5h9160wg2wd94-haskell-setenv-0.1.1.3.drv
  /nix/store/82hnbcdd415bndb9dalkf4a6b5rj89qn-haskell-hspec-meta-2.1.7.drv
  /nix/store/lp5d81c4kwjl7ssv3d901y13irm7rcj8-haskell-hspec-core-2.1.7.drv
  /nix/store/qx6rrzlkirdknb3r7l9xmkcskiykxd6f-haskell-hspec-discover-2.1.7.drv
  /nix/store/ad323c13gfkp5258rblwjmzjx2makn4c-haskell-hspec-2.1.7.drv
  /nix/store/drg2rif4gj2agf9350rf7vngyyy2ahha-haskell-stringbuilder-0.5.0.drv
  /nix/store/v2y2l56m8yhlry2kvhnmnbzr9xmjl17c-haskell-base-compat-0.8.1.1.drv
  /nix/store/y83myi25rvhmv8jlqbxhwbmxj7hwvl9z-haskell-ghc-paths-0.1.0.9.drv
  /nix/store/n6k2cavxcv7aq685wqz1pqiyai04aakm-haskell-doctest-0.9.13.drv
  /nix/store/k6w1fpmxi44z5f050dmbq9xx1bffcjcr-haskell-tagged-0.8.0.1.drv
  /nix/store/x77h6dad29k67lsn4ryffrr41z83khbl-haskell-distributive-0.4.4.drv
  /nix/store/sy0mrdmqjp5lapk38lg8rbdpj5bfp2ri-haskell-comonad-4.2.5.drv
  /nix/store/0hkkw0j04jf6r4bilxj7hi0h1qr5i699-haskell-semigroupoids-4.3.drv
  /nix/store/1lr5lhd7lgw0218nn8cdivcqdhhzhp12-haskell-unbounded-delays-0.1.0.9.drv
  /nix/store/z9lqh7fvizbnfy4m915hxhaw95iy10b0-haskell-parsec-3.1.9.drv
  /nix/store/7x1bzn5gncc1c5706si3p060n980hq2k-haskell-regex-tdfa-rc-1.1.8.3.drv
  /nix/store/jrpkxjg81lq4mjy7hwgzsij60hma4n66-haskell-optparse-applicative-0.11.0.2.drv
  /nix/store/0y6sxyw1rpy1xsww6bvh78bn2ndc6bd9-haskell-tasty-0.10.1.2.drv
  /nix/store/6zp7qh1cd4g2mkfiwz55d6rkh2f277aq-haskell-tasty-hunit-0.9.2.drv
  /nix/store/l1bi7ybac5l1kwx0rc34vvb43rcbyq7f-haskell-pcre-light-0.4.0.3.drv
  /nix/store/9c91gj73sqydph467ldzkd5rk896zfx4-haskell-tasty-quickcheck-0.8.3.2.drv
  /nix/store/gmcgq01qjvm75pgw0154370nak6sn9rv-haskell-utf8-string-0.3.8.drv
  /nix/store/gnfpiq5wlwsca42vv8w4j4r2ip6i37v5-haskell-language-haskell-extract-0.2.4.drv
  /nix/store/ywaf9hrkhsnair2k8i3i8wgm2w8ri98k-haskell-tasty-th-0.1.3.drv
  /nix/store/1yqda4a0qv3waplj2iaksh77ng06fyg0-haskell-multiarg-0.30.0.8.drv
  /nix/store/xrnvw4w9fxly5xmnm7ab6dxh20sdm0wa-haskell-logict-0.6.0.2.drv
  /nix/store/24hclwkycnb55qawh40yx6jpjs0193qb-haskell-smallcheck-1.1.1.drv
  /nix/store/2gxl6qyb14f5aksaznw1gz3j1vqamn71-haskell-old-time-1.1.0.3.drv
  /nix/store/2ifs8rrd8k87svbwmshydnqg57chw72j-haskell-generic-deriving-1.7.0.drv
  /nix/store/g9p701xkk28c4v3j3356sv3vzi7zb7dl-haskell-polyparse-1.11.drv
  /nix/store/44nv03h5rn6ly0i3jd56pgyl7p88s431-haskell-cpphs-1.19.drv
  /nix/store/5s9rfb9lyikk4jlw67xjfdhqz4f7y3mm-haskell-blaze-builder-0.3.3.4.drv
  /nix/store/6h7wzifyik3915y7m8rsin5r0rzr12v1-haskell-vector-0.10.12.3.drv
  /nix/store/88jfyvblj5dslb8yzf355rkgiwnx7pwg-haskell-tasty-ant-xml-1.0.1.drv
  /nix/store/i9gmr3vcalpndg2fbcxgdmcvckvvq8m3-haskell-tasty-smallcheck-0.8.0.1.drv
  /nix/store/b2ijg1a851hmjz5v47ks5yr5fr8yvc9i-haskell-scientific-0.3.3.8.drv
  /nix/store/72rgisq6bds7wrryxsqdw27vz2pdbj9z-haskell-attoparsec-0.13.0.0.drv
  /nix/store/45r64mn436yvczdx5c6vxmirgn3pqcdg-haskell-aeson-0.8.1.0.drv
  /nix/store/4fkn19xwaxw8xhc7jg2w4h1ihq23kvhm-haskell-exceptions-0.8.0.2.drv
  /nix/store/6sq09x1jga9lmknwnx2kx2r17sq9dfa7-haskell-cmdargs-0.10.12.drv
  /nix/store/ajkbmnmi79l2vkwwz351ihk0kxahg4wv-haskell-profunctors-4.4.1.drv
  /nix/store/g0qwgczk2xpxifq86fwhvkingdbkkdv3-haskell-prelude-extras-0.4.drv
  /nix/store/vcw40ai9nwm7003ma758b95xqkrlv1v1-haskell-bifunctors-4.2.1.drv
  /nix/store/hycbsh8d933gi7g8q092ywsqn0b45zg4-haskell-free-4.11.drv
  /nix/store/954bln9rjng5wjl03mk8k64qd6baww47-haskell-adjunctions-4.2.drv
  /nix/store/8sn4960dnn88vc1rhsijg11ckj4lrbrg-haskell-kan-extensions-4.2.1.drv
  /nix/store/i4535n1khcwb82vi7lsf2q8wv6sslav9-haskell-temporary-rc-1.2.0.3.drv
  /nix/store/9i0b98rjl5plgdspc7jzwb7c77y4jc5b-haskell-tasty-golden-2.3.0.1.drv
  /nix/store/i56l80yzahz7z0k8ffj38zx2baylyzg0-haskell-simple-reflect-0.3.2.drv
  /nix/store/cn2sinl7lbkzdhkyb6m76qh2nm2lwjps-haskell-uniplate-1.6.12.drv
  /nix/store/fh77rwzcldixka1kj5gv8z1yidwbj5i4-haskell-hscolour-1.23.drv
  /nix/store/rcjkl8mr55i37k21z3ihib8ksqkdlkyi-haskell-haskell-src-exts-1.16.0.1.drv
  /nix/store/v7pcfg8faiyq6jx5mvjhvp6gniwyymid-haskell-extra-1.1.drv
  /nix/store/iq8cijd03hmypz3phbpxwymxqb55sf3p-haskell-hlint-1.9.20.drv
  /nix/store/mhcshqpyh4q90x80hj1j9qs0sy748v7s-haskell-reflection-1.5.1.2.drv
  /nix/store/nfb8v3i5lxkiyhrc9lg1vjagmn4nvxd4-haskell-test-framework-th-0.2.4.drv
  /nix/store/v8pkvi4frk9r28bcbnhddrxf1f5qf5km-haskell-parallel-3.2.0.6.drv
  /nix/store/an37xgg2j1mar48kr3vhdcbq29n0n5vx-haskell-lens-4.9.1.drv
  /nix/store/bkf0bf7j3b8gr8h0wk465vczsaackgz8-haskell-mwc-random-0.13.3.2.drv
  /nix/store/by3y250m4pj4nnadk8r7z1qz69l91iab-haskell-SHA-1.6.4.2.drv
  /nix/store/qjza89991pzbr7brx0k0r3rphriwlll8-haskell-tar-0.4.1.0.drv
  /nix/store/dvcd7fz5h5855vmd4vibyvrfzn1vy67y-haskell-hackage-db-1.22.drv
  /nix/store/hqw8clsn2d62pc54iybi61nxyi07hbhz-haskell-split-0.2.2.drv
  /nix/store/kmy6ywpr2wjggmv7hyva0x6y98gay5d7-haskell-abstract-deque-0.3.drv
  /nix/store/x48am4xy5a9av2csza43glhl4xy22qy1-haskell-abstract-par-0.3.3.drv
  /nix/store/xl1ffw77zd55cg0xp0jnl1g8qavvk5k7-haskell-cereal-0.4.1.1.drv
  /nix/store/lhhv6a2bjhvbddxxyzx0cw3xykx36m7h-haskell-monad-par-extras-0.3.3.drv
  /nix/store/pyl9zcr00k7hzh0ljzwb1myd7wxp9p5p-haskell-monad-par-0.3.4.7.drv
  /nix/store/rlzccilkgmpvniv509jpzyjwzsrhg0a1-haskell-deepseq-generics-0.1.1.2.drv
  /nix/store/z039zp2l8p34wxrlah9hszrjjgxbmypv-haskell-cartel-0.14.2.6.drv
  /nix/store/xqy54j75y81nanr4fzb3blniafjl9mh4-cabal2nix-20150505.drv
building path(s) ‘/nix/store/ygp6s44nxv4z7am2g8g52iwqkbgxj04h-ghc-7.4.2-binary’
unpacking sources
unpacking source archive /nix/store/4zj30pw1rzvqshnrcsl9dw1ahdxhlz0z-ghc-7.4.2-i386-unknown-linux.tar.bz2
source root is ghc-7.4.2
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
missing ELF header
not an ELF executable
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
cannot find section .interp
not an ELF executable
not an ELF executable
patching sources
configuring
checking for path to top of build tree... /tmp/nix-build-ghc-7.4.2-binary.drv-0/ghc-7.4.2
ghc/stage2/build/tmp/ghc-stage2: error while loading shared libraries: libncurses.so.5: cannot open shared object file: No such file or directory
Can't work out build platform
builder for ‘/nix/store/vawns4div0z5cq1w1wvch0jgw7zq40fj-ghc-7.4.2-binary.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/vdgwwldb3a8fwv9bvmrc21cqqmb0zidl-ghc-7.8.4.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/68pkhyfymp5vrbysyzgrsbd621gxvqyw-ghc-7.10.1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/xqy54j75y81nanr4fzb3blniafjl9mh4-cabal2nix-20150505.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/xqy54j75y81nanr4fzb3blniafjl9mh4-cabal2nix-20150505.drv’ failed
```

**Comment by peti** [comment]
@wkennington, ping?

**Comment by Warbo** [comment]
FYI I get the same error installing with nix-env and nixos-rebuild from the unstable channel, and with the latest git commit ( 8a53e1a )

**Comment by vcunat** [comment]
@Warbo: it hasn't been solved yet on any branch AFAIK.

**Comment by wkennington** [comment]
I took a look but I have no idea how this is broken yet.

**Comment by Warbo** [comment]
I ran `git checkout 41b53577a8f2 -- pkgs/development/libraries/ncurses` and both cabal2nix and cabal-install worked.

Prior to that, I also tried naively making libncurses.so.5 a symlink to libncurses.so at the end of the postInstall, but that gave a "file too short" error.

**Comment by peti** [comment]
IMHO, this error is rather serious. I would suggest reverting the `ncurses` refactoring until a fix has been found.

**Comment by edolstra** [comment]
I assume the problem is that this line got dropped by ee78173b213059e6c885dc0c4d3bf4da7d3cb229:
```
        ln -svf lib''${lib}w.so.5 $out/lib/lib$lib.so.5
```
I'm also in favor of reverting ee78173b213059e6c885dc0c4d3bf4da7d3cb229, also because I don't really like the `configureFlags` handling style introduced in that commit.

**Comment by wkennington** [comment]
Should be fixed by 5a89a87cc575f0d8b9f6cc20e8203c18ac30a8c0

**Comment by peti** [comment]
@wkennington, thank you for fixing this issue!
